### ISSUE TITLE
Bubble opacity is not reset during series.update

### ISSUE THREAD

**Issue by kzoon** [highcharts/highcharts#12036]
#### Expected behaviour
The series is 'forgets' its options of before the update, as documented in https://api.highcharts.com/class-reference/Highcharts.Series#update 

So, if the series has an options for marker.fillopacity and gets update without an options for marker.fillopacity, the opacitoy of the bubble reverts to the value set in plotOptions

#### Actual behaviour
The marker.fillopacity is not removed

#### Live demo with steps to reproduce
In https://jsfiddle.net/kzoon/oaesyudt/ the blue series should become semi-transparent after 'update opacity' is clickecd

**Comment by KacperMadej** [comment]
Hi @kzoon 

Thank you for reporting the problem.

This is a general issue with updates - there is no value that would clear the option on update. If you will set `undefined`, then it will be used instead of falling back to plotOptions value or default series value.

As a workaround, you could clear new option in series `init` event if it will be updated to something custom, let's say `'delete'` - demo: https://jsfiddle.net/BlackLabel/u8ewb6ha/

```js
      	init: function(obj) {
      	      	if (obj.options.marker && obj.options.marker.fillOpacity === 'delete') {
      	      	      	delete obj.options.marker.fillOpacity;
      	      	      	console.log('Worked around!');
      	      	}
      	}
```

As a more general solution, you could merge default options manually when you want to update a series - demo: https://jsfiddle.net/BlackLabel/u8ewb6ha/1/

**Comment by pawelfus** [comment]
I don't know.. default value for `fillOpacity` is `0.5`. User option is `1` so why **no setting** of an option should clear it? I think it works the most intuitive way: `series.update()` updates series with new options. Using `series.update({ data })` updates data leaving all other options untouched. 

> The series is 'forgets' its options of before the update, as documented in

This is wrong. In docs we say:

> For a clean and precise handling of new options, all **methods** and **elements** from the series are removed, and it is initialized from scratch. Therefore, this method is more performance expensive than some other utility methods (...)

That elements like dataLabels, markers, graphs, lines and zones etc are created from scratch, reinitialized. It has nothing to do with forgetting initial options. By methods we mean, for example `series.addPoint()`, `series.update()` and all internal methods, that those methods are recreated too. That's why `series.update({ data })` is slower than `series.setData(data)` and we recommend to use `setData()` in this particular case.

In other words, to reset value, set it back to the default value explicitly: 
```js
  chart.series[0].update({
    name: 'series A',
    data: [
      [10, 10, 1],
      [20, 20, 2]
    ],
    marker: {
      fillOpacity: 0.5
    }
  });
```

Demo: https://jsfiddle.net/BlackLabel/L0mfhrpg/

**Comment by kzoon** [comment]
I disagree.
Maybe leaving out fillOpacity in the update options shouldn't work, but supplying a null value _should_.
You _can_ reset the color of a the marker on an individual series by assigning null, like so: https://jsfiddle.net/kzoon/qjmpyxw4/1/
If you try this with fillOpacity, the color turns to black. Even the second updated series goes black. See https://jsfiddle.net/kzoon/oaesyudt/5/
This can't be intended behavior.

**Comment by pawelfus** [comment]
I didn't mention `null` value in my comment above @kzoon. I agree with you, setting `fillOpacity: null` should restore default value. Just to be on the same page, so we all agree:

1. Providing explicit value (works):
```js
series.update({ marker: { fillOpacity: <number> } })
```

2. Not providing an option **does nothing** (works):
```js
series.update({ })
```

3. Providing **null** value should reset to default value (needs implementation):
```js
series.update({ marker: { fillOpacity: null } })
```

PS: Regarding second series going black - that's expected, compare colors for second and third series: https://jsfiddle.net/BlackLabel/ejzxk1g7/ :)

**Comment by kzoon** [comment]
I agree with your summing up.

**Comment by TorsteinHonsi** [comment]
Sorry for barging in, but I don't agree to the summary. The `fillOpacity` only accepts numbers, which is nice and type safe. By extension, if we should allow this concept all over our API, we essentially make all our option loosely typed. There is nothing separating the `fillOpacity` from other options in this regard. For example, `legend.enabled` is `true` by default. So if we implemented this concept, it would be possible to set it to `false`, then in the second run to `null` to restore it. That would mean that all our options would have typing like `boolean|null`, `number|null`, `string|null`, right?

@bre1470 FYI.

**Comment by bre1470** [comment]
We should avoid `null` as much as possible as we are targeting strict typing. For default behaviour resetting to `undefined` should be enough. If this is not working, then something is wrong with the internal merge logic.

**Comment by bre1470** [comment]
So instead of:

```ts
series.update({ marker: { fillOpacity: null } })
```
this should be possible:
```ts
series.update({ marker: { fillOpacity: void 0 /* undefined */ } })
```

**Comment by pawelfus** [comment]
Right, `undefined` is definitively better.

We still can use `Highcharts.getOptions()` to get default values: https://jsfiddle.net/BlackLabel/L0mfhrpg/1/

```js
  chart.series[0].update({
    name: 'series A',
    data: [
      [10, 10, 1],
      [20, 20, 2]
    ],
    marker: {
      fillOpacity: Highcharts.getOptions().plotOptions.bubble.marker.fillOpacity
    }
  });
```

**Comment by TorsteinHonsi** [comment]
But allowing setting it to `undefined` would also mean looser typing than currently, because currently it is (or should be) a `number`. Its default value is set in the default options, and it is always defined. Basically the argument is as the same as I made for `null` yesterday. Why allow `legend.enabled` to be `boolean|undefined` when it can be simply `boolean`?

I'm trying to come up with a good analogy in native JS - take for example the `style.width` of an element. According to [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/width) it can take a defined set of length units or keywords. Setting it to `undefined` does nothing: https://jsfiddle.net/highcharts/7t3kmpgj/4/.

**Comment by bre1470** [comment]
So what is the conclusion for this issue? Should we introduce a `delete()` function and ignore `null`/ `undefined` in `update()`?

**Comment by TorsteinHonsi** [comment]
No, we don't need that, we can reset the `fillOpacity` by setting it back to its default value. Like with any other option.

**Comment by kzoon** [comment]
This is one of the bugs that contributes to the enormous amount of workaround code in my charting code. Look like there is a solution waiting to be implemented.
